portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -136,10 +215,41 @@ private PreparedQuery getAllComments() {
     return datastore.prepare(query);
   }
 
+  /** 
+    * Returns all user comments that contain `search` in either
+    * the comment text or author name. 
+    * @param results All available comment entities in the database.
+    * @param comments An empty list to be filled. 
+    * @param search A string to filter the comments by. 
+    */
+  private ArrayList<Comment> getFilteredComments(PreparedQuery results, 
+    ArrayList<Comment> comments, String search) {
+    results.asIterable().forEach(commentEntity -> {
+      comments.add(Comment.fromEntity(commentEntity));
+    });
+
+    ArrayList<Comment> filteredComments = new ArrayList<Comment>();
+    Iterables.filter(comments, c -> satisfiesSearch(c, search))
+      .forEach(c -> filteredComments.add(c));
+    return filteredComments;
+  }
+
+  /** 
+    * Returns true if this comment contains the search string in the
+    * comment text or author name. 
+    */
+  private boolean satisfiesSearch(Comment comment, String search) {
+    if (Strings.isNullOrEmpty(search)) {
+      return true;
+    } else {
+      return comment.text.contains(search) || comment.name.contains(search);
+    }
+  }
+
   /** Returns JSON string representation of `data`. */
-  private String convertToJson(List<Comment> data) {
+  private String convertToJson(List<Comment> data, int pageToken) {
     Gson gson = new Gson(); 
     String json = gson.toJson(data);
-    return json;
+    return "[{\"pageToken\" : "+pageToken+" }, "+json+"]";
```

Comment:
It's important to consider the cost of String addition. (though fine for this project as a starter).  


This implementation would cost O(n) for making the original json, then at worst, due to 4 additions, it could be copied 4 times. 

Instead, StringBuilder is a good way to go.

Reply:
Wow I had never heard of this class! Great idea! fixed.